## 4.1 Problem Definition
**Selected Dataset:** PlantVillage Dataset (from Kaggle)  
**Dataset Source:** [Kaggle - PlantVillage Dataset](https://www.kaggle.com/datasets/abdallahalidev/plantvillage-dataset)

### Project Overview:
This dataset contains over 54,000 images of healthy and diseased crop leaves across several species. In agriculture, early detection of plant diseases is crucial for food security and reducing pesticide use. This project represents a data investigation into how these diseases are distributed and analyzed within a large-scale image dataset.

### Analytical Questions:
1. **Distribution Analysis:** Which crop species (e.g., Apple, Tomato) are the most and least represented in this dataset? 
   * *Why it matters:* This helps identify if the dataset is biased toward specific crops, which affects how reliable a disease detection model might be for other crops.
2. **Anomaly Detection (NumPy):** Are there any unusual image filenames (outliers) that might indicate data corruption or inconsistent naming conventions? 
   * *Why it matters:* Finding outliers ensures data quality and helps in cleaning before feeding data into a machine learning model.
3. **Relationship Analysis:** Does the variety of diseases found in a specific crop correlate with the total number of images available for that crop?
   * *Why it matters:* To understand if more data leads to discovering more disease types or if certain crops are naturally more prone to diseases.

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

In [13]:
import os
import pandas as pd
import numpy as np


base_path = r'D:\lectures slide\10th semester\Python\Final\Final_Project\dataset\plantvillage dataset\color'

if os.path.exists(base_path):
    all_data = []
    
    for folder in os.listdir(base_path):
        folder_path = os.path.join(base_path, folder)
        
        if os.path.isdir(folder_path):
            for file in os.listdir(folder_path):
                if file.lower().endswith(('.png', '.jpg', '.jpeg')):
                    all_data.append({
                        'Image_Name': file,
                        'Disease_Type': folder
                    })
    
    
    df = pd.DataFrame(all_data)
    print("Successfully loaded data!")
    print(f"Total Unique Image: {len(df)}")
    print("\n1st 5 row:")
    print(df.head())
else:
    print("path not found! kindly check folder path")

Successfully loaded data!
Total Unique Image: 54305

1st 5 row:
                                          Image_Name        Disease_Type
0  00075aa8-d81a-4184-8541-b692b78d398a___FREC_Sc...  Apple___Apple_scab
1  01a66316-0e98-4d3b-a56f-d78752cd043f___FREC_Sc...  Apple___Apple_scab
2  01f3deaa-6143-4b6c-9c22-620a46d8be04___FREC_Sc...  Apple___Apple_scab
3  0208f4eb-45a4-4399-904e-989ac2c6257c___FREC_Sc...  Apple___Apple_scab
4  023123cb-7b69-4c9f-a521-766d7c8543bb___FREC_Sc...  Apple___Apple_scab


In [14]:
df.sample(10)

,Image_Name,Disease_Type
14870,108f2813-6225-4172-9b14-3e45f0a84138___UF.Citr...,Orange___Haunglongbing_(Citrus_greening)
28984,4492489a-315f-4eca-9e43-95762f3044d1___RS_HL 5...,Soybean___healthy
5990,6f863320-0323-490f-a02d-b943a9caa94f___FREC_Pw...,Cherry_(including_sour)___Powdery_mildew
3772,6b9195b3-327a-473b-a455-4a77b1d6d2ed___RS_HL 0...,Blueberry___healthy
43217,7f9fdc98-b176-45ac-a56f-28accf0a5a95___Crnl_L....,Tomato___Leaf_Mold
28107,1770ce1e-e269-48c7-b7eb-4e19ddd0ab55___RS_HL 3...,Soybean___healthy
36578,371929a8-17b8-4077-94f4-cfb6963d9285___UF.GRC_...,Tomato___Bacterial_spot
37963,dec2db16-8bae-42e0-9b02-8e3d9f0793cc___GCREC_B...,Tomato___Bacterial_spot
13514,19609816-c12d-4f2a-a6a8-b55471f35436___FAM_L.B...,Grape___Leaf_blight_(Isariopsis_Leaf_Spot)
19558,ec46c5d4-849e-4b7c-9b30-65d8c8c3bd8e___CREC_HL...,Orange___Haunglongbing_(Citrus_greening)


In [15]:
disease_counts = df['Disease_Type'].value_counts()
print(disease_counts)

Disease_Type
Orange___Haunglongbing_(Citrus_greening)              5507
Tomato___Tomato_Yellow_Leaf_Curl_Virus                5357
Soybean___healthy                                     5090
Peach___Bacterial_spot                                2297
Tomato___Bacterial_spot                               2127
Tomato___Late_blight                                  1909
Squash___Powdery_mildew                               1835
Tomato___Septoria_leaf_spot                           1771
Tomato___Spider_mites Two-spotted_spider_mite         1676
Apple___healthy                                       1645
Tomato___healthy                                      1591
Blueberry___healthy                                   1502
Pepper,_bell___healthy                                1478
Tomato___Target_Spot                                  1404
Grape___Esca_(Black_Measles)                          1383
Corn_(maize)___Common_rust_                           1192
Grape___Black_rot                          

In [16]:
df.tail()

,Image_Name,Disease_Type
54300,ffb295c9-f14e-4a15-831a-bf905da7fcb6___UF.GRC_...,Tomato___Tomato_Yellow_Leaf_Curl_Virus
54301,ffe08ccc-c55e-4ca2-9234-2906b98b8d05___YLCV_NR...,Tomato___Tomato_Yellow_Leaf_Curl_Virus
54302,ffe996e5-c8dc-47b7-bca2-4fc25e5ac57c___UF.GRC_...,Tomato___Tomato_Yellow_Leaf_Curl_Virus
54303,fff42f1b-7ec4-46e3-9269-45932e63635e___YLCV_GC...,Tomato___Tomato_Yellow_Leaf_Curl_Virus
54304,fff6bf83-797c-4de4-83f7-39b74786fec4___UF.GRC_...,Tomato___Tomato_Yellow_Leaf_Curl_Virus


## 1. Problem Definition
*   **Dataset:** PlantVillage Dataset (Kaggle)
*   **Topic:** This dataset represents various healthy and diseased plant leaves.
*   **Analytical Questions:**
    1. Which plant species/diseases have the highest and lowest representation in the dataset? (Distribution Analysis)
    2. How does the average image filename length vary across different disease categories? (Numerical/Metadata Analysis)
    3. Are there any specific crops (e.g., Tomato, Apple) that are more prone to a wider variety of diseases in this dataset?

In [11]:
#  Data Understanding
print(f"Dataset Shape: {df.shape}")
print(f"Columns: {df.columns.tolist()}")

#  Data Cleaning

initial_count = len(df)
df = df.drop_duplicates()
print(f"Justification: Removed {initial_count - len(df)} duplicate entries to ensure unique image analysis.")


print("\nMissing Values:\n", df.isnull().sum())
print("Justification: Verified there are no null values to avoid errors during aggregation.")


df['Disease_Type'] = df['Disease_Type'].str.replace('___', ' | ')
print("Justification: Replaced '___' with '|' for better readability in visualizations.")

df.head()

Dataset Shape: (54305, 2)
Columns: ['Image_Name', 'Disease_Type']
Justification: Removed 0 duplicate entries to ensure unique image analysis.

Missing Values:
 Image_Name      0
Disease_Type    0
dtype: int64
Justification: Verified there are no null values to avoid errors during aggregation.
Justification: Replaced '___' with '|' for better readability in visualizations.


,Image_Name,Disease_Type
0,00075aa8-d81a-4184-8541-b692b78d398a___FREC_Sc...,Apple | Apple_scab
1,01a66316-0e98-4d3b-a56f-d78752cd043f___FREC_Sc...,Apple | Apple_scab
2,01f3deaa-6143-4b6c-9c22-620a46d8be04___FREC_Sc...,Apple | Apple_scab
3,0208f4eb-45a4-4399-904e-989ac2c6257c___FREC_Sc...,Apple | Apple_scab
4,023123cb-7b69-4c9f-a521-766d7c8543bb___FREC_Sc...,Apple | Apple_scab


##  Data Understanding - Interpretation
Based on the code outputs above, we can summarize the dataset structure as follows:

*   **Dataset Shape:** The dataset contains **54,305 rows** and **2 initial columns**, meeting the requirement of at least 1,000 rows.
*   **Column Structure:** 
    *   `Image_Name`: A categorical/string column containing the unique identifiers for each leaf image.
    *   `Disease_Type`: A categorical column representing the crop name and the specific health status (healthy or diseased).
*   **Data Quality:** 
    *   **Missing Values:** The summary shows zero null values across all columns, indicating a high-quality, complete dataset.
    *   **Duplicates:** Our cleaning step confirmed and removed any redundant entries, ensuring the analytical findings are based on unique evidence.

In [26]:

df['Crop'] = df['Disease_Type'].apply(lambda x: x.split('___')[0])


df['Health_Status'] = df['Disease_Type'].apply(lambda x: 'Healthy' if 'healthy' in x.lower() else 'Diseased')


print("Total Unique Crops now:", df['Crop'].nunique())
print("Crop list:", df['Crop'].unique())
df.head()

Total Unique Crops now: 14
Crop list: <StringArray>
[                  'Apple',               'Blueberry',
 'Cherry_(including_sour)',            'Corn_(maize)',
                   'Grape',                  'Orange',
                   'Peach',            'Pepper,_bell',
                  'Potato',               'Raspberry',
                 'Soybean',                  'Squash',
              'Strawberry',                  'Tomato']
Length: 14, dtype: str


,Image_Name,Disease_Type,Crop,Health_Status,Name_Length,Z_Score
0,00075aa8-d81a-4184-8541-b692b78d398a___FREC_Sc...,Apple___Apple_scab,Apple,Diseased,57,0.067749
1,01a66316-0e98-4d3b-a56f-d78752cd043f___FREC_Sc...,Apple___Apple_scab,Apple,Diseased,57,0.067749
2,01f3deaa-6143-4b6c-9c22-620a46d8be04___FREC_Sc...,Apple___Apple_scab,Apple,Diseased,57,0.067749
3,0208f4eb-45a4-4399-904e-989ac2c6257c___FREC_Sc...,Apple___Apple_scab,Apple,Diseased,57,0.067749
4,023123cb-7b69-4c9f-a521-766d7c8543bb___FREC_Sc...,Apple___Apple_scab,Apple,Diseased,57,0.067749


In [27]:
healthy_samples = df[df['Health_Status'] == 'Healthy']
print(f"Total Healthy Images: {len(healthy_samples)}")
display(healthy_samples.head(10))

Total Healthy Images: 15084


,Image_Name,Disease_Type,Crop,Health_Status,Name_Length,Z_Score
1526,0055dd26-23a7-4415-ac61-e0b44ebfaf80___RS_HL 5...,Apple___healthy,Apple,Healthy,53,-0.505408
1527,00907d8b-6ae6-4306-bfd7-d54471981a86___RS_HL 5...,Apple___healthy,Apple,Healthy,53,-0.505408
1528,0098dbd9-286a-4d6a-bf4b-5459d66f88c0___RS_HL 5...,Apple___healthy,Apple,Healthy,53,-0.505408
1529,00a6039c-e425-4f7d-81b1-d6b0e668517e___RS_HL 7...,Apple___healthy,Apple,Healthy,53,-0.505408
1530,00fca0da-2db3-481b-b98a-9b67bb7b105c___RS_HL 7...,Apple___healthy,Apple,Healthy,53,-0.505408
1531,010125c0-e6f2-4f3d-b469-dd98ffecbb01___RS_HL 7...,Apple___healthy,Apple,Healthy,53,-0.505408
1532,011d02f3-5c3c-4484-a384-b1a0a0dbdec1___RS_HL 7...,Apple___healthy,Apple,Healthy,53,-0.505408
1533,013b7c70-5e3b-42b7-86af-167815a5b04f___RS_HL 7...,Apple___healthy,Apple,Healthy,53,-0.505408
1534,017fd21b-142f-4988-849a-5fd316ea9a1b___RS_HL 6...,Apple___healthy,Apple,Healthy,53,-0.505408
1535,0181db4e-718e-41bc-b6ed-92863471fa85___RS_HL 5...,Apple___healthy,Apple,Healthy,53,-0.505408


##  Feature Engineering
In this section, we created 3 new derived columns from the existing data to support deeper analysis. This moves beyond the original metadata to provide more granular insights.

### New Features Created:
*   **`Crop`**: Extracted from `Disease_Type`. It allows us to perform subgroup comparisons by plant species (e.g., Apple vs. Tomato).
*   **`Health_Status`**: A categorical feature derived by checking for the keyword 'healthy' in the disease name. This classifies each image as either **Healthy** or **Diseased**.
*   **`Name_Length`**: A numerical feature representing the character count of each image filename. This is essential for performing statistical analysis and outlier detection using NumPy.

### Why these features matter:
These features allow us to answer complex questions, such as the ratio of healthy to diseased plants and whether certain filename patterns correlate with specific categories.

In [28]:
import numpy as np

crop_health_comparison = df.groupby(['Crop', 'Health_Status']).size().unstack().fillna(0)
print("--- Crop wise Health Status Comparison ---")
print(crop_health_comparison)

--- Crop wise Health Status Comparison ---
Health_Status            Diseased  Healthy
Crop                                      
Apple                      1526.0   1645.0
Blueberry                     0.0   1502.0
Cherry_(including_sour)    1052.0    854.0
Corn_(maize)               2690.0   1162.0
Grape                      3639.0    423.0
Orange                     5507.0      0.0
Peach                      2297.0    360.0
Pepper,_bell                997.0   1478.0
Potato                     2000.0    152.0
Raspberry                     0.0    371.0
Soybean                       0.0   5090.0
Squash                     1835.0      0.0
Strawberry                 1109.0    456.0
Tomato                    16569.0   1591.0


In [29]:
disease_variety = df.groupby('Crop')['Disease_Type'].nunique().sort_values(ascending=False)
print("\n--- Ranking: Crops with Most Disease Varieties ---")
print(disease_variety)


--- Ranking: Crops with Most Disease Varieties ---
Crop
Tomato                     10
Apple                       4
Grape                       4
Corn_(maize)                4
Potato                      3
Cherry_(including_sour)     2
Strawberry                  2
Peach                       2
Pepper,_bell                2
Blueberry                   1
Orange                      1
Raspberry                   1
Squash                      1
Soybean                     1
Name: Disease_Type, dtype: int64


In [30]:
name_lengths = df['Name_Length'].values
mean_len = np.mean(name_lengths)
std_len = np.std(name_lengths)
print(f"\n--- NumPy Statistics ---")
print(f"Mean Filename Length: {mean_len:.2f}")
print(f"Standard Deviation: {std_len:.2f}")


--- NumPy Statistics ---
Mean Filename Length: 56.53
Standard Deviation: 6.98


In [31]:
df['Z_Score'] = (df['Name_Length'] - mean_len) / std_len
outliers = df[np.abs(df['Z_Score']) > 3]
print(f"\n--- Outlier Analysis ---")
print(f"Total Anomalies Found (Z-Score > 3): {len(outliers)}")


--- Outlier Analysis ---
Total Anomalies Found (Z-Score > 3): 1193


In [32]:
relationship = df.groupby('Crop').agg({
    'Image_Name': 'count', 
    'Disease_Type': 'nunique'
}).rename(columns={'Image_Name': 'Total_Images', 'Disease_Type': 'Unique_Diseases'})
print("\n--- Relationship: Data Volume vs Disease Variety ---")
print(relationship)


--- Relationship: Data Volume vs Disease Variety ---
                         Total_Images  Unique_Diseases
Crop                                                  
Apple                            3171                4
Blueberry                        1502                1
Cherry_(including_sour)          1906                2
Corn_(maize)                     3852                4
Grape                            4062                4
Orange                           5507                1
Peach                            2657                2
Pepper,_bell                     2475                2
Potato                           2152                3
Raspberry                         371                1
Soybean                          5090                1
Squash                           1835                1
Strawberry                       1565                2
Tomato                          18160               10


In [33]:
print("Total Unique Crops:", df['Crop'].nunique())
print("List of Crops:", df['Crop'].unique())

Total Unique Crops: 14
List of Crops: <StringArray>
[                  'Apple',               'Blueberry',
 'Cherry_(including_sour)',            'Corn_(maize)',
                   'Grape',                  'Orange',
                   'Peach',            'Pepper,_bell',
                  'Potato',               'Raspberry',
                 'Soybean',                  'Squash',
              'Strawberry',                  'Tomato']
Length: 14, dtype: str


##  Data Analysis
In this section, we performed 6 major analytical operations as per the project requirements:

1.  **Subgroup Comparison:** We compared the frequency of healthy and diseased leaves across different crops using grouping and unstacking.
2.  **Ranking Analysis:** Identified which crop species has the widest variety of diseases in this dataset (e.g., Tomato usually ranks high).
3.  **Custom NumPy Computation:** Utilized `np.mean()` and `np.std()` to calculate statistical metrics for filename lengths.
4.  **Outlier/Anomaly Detection:** Calculated **Z-Scores** to identify filenames that are statistically unusual (more than 3 standard deviations from the mean).
5.  **Relationship Analysis:** Evaluated the correlation between the total number of images collected for a crop and the number of unique diseases identified for that crop.
6.  **Frequency Analysis:** Summarized the overall health status distribution across the entire dataset.

In [2]:
%pip install pandas numpy matplotlib

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 26.1
[notice] To update, run: python.exe -m pip install --upgrade pip
